python/pyarrow/lib_gpu.pyx

```diff
@@ -0,0 +1,809 @@
+        return pyarrow_wrap_cudacontext(self.cuda_buffer.get().context())
+
+    def slice(self, offset=0, length=None):
+        """Return slice of device buffer
+
+        Parameters
+        ----------
+        offset : int, default 0
+          Specify offset from the start of device buffer to slice
+        length : int, default None
+          Specify the length of slice (default is until end of device
+          buffer starting from offset)
+
+        Returns
+        -------
+        sliced : CudaBuffer
+          Zero-copy slice of device buffer.
+        """
+        if offset < 0 or offset >= self.size:
+            raise ValueError('offset argument is out-of-range')
+        cdef int64_t offset_ = offset
+        cdef int64_t size
+        if length is None:
+            size = self.size - offset_
+        elif offset + length <= self.size:
+            size = length
+        else:
+            raise ValueError(
+                'requested larger slice than available in device buffer')
+        parent = pyarrow_unwrap_cudabuffer(self)
+        return pyarrow_wrap_cudabuffer(
+            shared_ptr[CCudaBuffer](make_shared[CCudaBuffer](parent,
+                                                             offset_, size)))
+
+    def to_pybytes(self):
+        return self.copy_to_host().to_pybytes()
+
+    def __getbuffer__(self, cp.Py_buffer * buffer, int flags):
+        # copy_to_host() is needed otherwise, the buffered device
+        # buffer would contain data pointers of the device
+        raise NotImplementedError('CudaBuffer.__getbuffer__')
+
+    def __getreadbuffer__(self, Py_ssize_t idx, void ** p):
+        raise NotImplementedError('CudaBuffer.__getreadbuffer__')
+
+    def __getwritebuffer__(self, Py_ssize_t idx, void ** p):
+        raise NotImplementedError('CudaBuffer.__getwritebuffer__')
+
+
+cdef class CudaHostBuffer(Buffer):
+    """Device-accessible CPU memory created using cudaHostAlloc.
+
+    To create a CudaHostBuffer instance, use
+
+      <CudaDeviceManager instance>.allocate_host(<nbytes>)
+
+    To free memory allocated in CudaHostBuffer instance, use
+
+      <CudaDeviceManager instance>.free_host(<CudaHostBuffer instance>)
+
+    The memory is also automatically freed when CudaHostBuffer
+    instance is deleted.
+
+    Note: after freeing CudaHostBuffer memory, the instance should be
+    discarded as unusable.
+    """
+
+    def __cinit__(self):
+        self._freed = True
+
+    def __init__(self):
+        raise TypeError("Do not call CudaHostBuffer's constructor directly,"
+                        " use `allocate_cuda_host_buffer` function instead.")
+
+    cdef void init_host(self, const shared_ptr[CCudaHostBuffer] & buffer):
+        self.host_buffer = buffer
+        self.init(< shared_ptr[CBuffer] > buffer)
+        self._freed = False
+
+    @property
+    def size(self):
+        if self._freed:
+            return 0
+        return self.host_buffer.get().size()
+
+    def __dealloc__(self):
```

Comment:
This should be removed. The C++ layer is responsible for freeing resources. Even if the Python object disappears, the C++ `shared_ptr` may still be used elsewhere (or be wrapped in other Python objects).

Reply:
Done. Removed FreeHost from pyarrow.cuda.